src/dmd/root/outbuffer.d

```diff
@@ -25,9 +25,12 @@ debug
 
 struct OutBuffer
 {
+    import dmd.root.file : FileMapping;
+
     private ubyte[] data;
     private size_t offset;
     private bool notlinehead;
+    private FileMapping!ubyte fileMapping;
```

Comment:
What is happening here is that the code in OutBuffer is mostly reused, and intermingled here and there are exceptions for FileMapping. This sounds like a perfect use case for inheritance.

I recommend making a new struct, FileMapBuffer, that has a field of type OutBuffer. Set `alias this` to point to the OutBuffer field. Then override the bits that are different for FileMapBuffer. I did the same thing with RArray inheriting from BArray in https://github.com/dlang/dmd/blob/master/src/dmd/backend/barray.d#L221

Reply:
What is someone gets the reference to the `OutBuffer` and then calls `resize()` against it?

Reply:
> What is someone gets the reference to the `OutBuffer` and then calls `resize()` against it?

They'll be a compiler error because OutBuffer doesn't implement `resize()`?

If I understand your concern correctly, care should be taken to avoid passing the interior `OutBuffer` around. Making it `private` would prevent that though.

Reply:
I'm confused. If `FileMapBuffer` is a subtype of `OutBuffer` (via alias this) then a reference to `FileMapBuffer` converts implicitly to a reference to `OutBuffer`. (So can't make that thing private.) That means people would use an `OutBuffer` that in effect would transparently do memory-mapped stuff underneath when resizing. You can't do that without virtual functions, but I hope we agree that's overkill.

Not to mention mmfiles can be used for reading files (see the comparison in `update`), so it would seem odd that I use a subtype of `OutBuffer` for reading.

As things are right now: `FileMapping` is an abstraction that takes care of all the nonsense associated with memory-mapped files. It's not for input or output and has nothing to do with `OutBuffer`.  In turn, `OutBuffer` may occasionally make use of said abstraction. I don't see why we need to (or actually can) use subtyping.

Reply:
I looked over https://github.com/dlang/dmd/blob/master/src/dmd/backend/barray.d#L221 and BArray is problematic.

It has a built-in array and an extra member `capacity`. Also it allows implicit conversion to `ref T[]` by means of `alias this`, effectively subtyping the built-in slice.

If there's any synchronization needed between `array` and `capacity`, the design is unsound because user code can get it unglued without a cast in sight. All they need to do is change the built-in slice stored inside the `BArray`, and then the capacity goes out of sync.

Structs and classes that can be only used in specific ways and cannot be verified independently are an endemic problem in dmd.